### PR TITLE
Feat/proof v1.2

### PR DESCRIPTION
**Pull request by mabmayer**
Proof for v1.2, refactoring

### CODE REVIEW

**Review by SimoneLazzaris — APPROVED** [review]
Too big to check every line, but overall LGTM

File: immudb/database/meta.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
+# Copyright 2021 CodeNotary, Inc. All rights reserved.
```
Comment (byo) [inline immudb/database/meta.py:2]:
```suggestion
# Copyright 2022 CodeNotary, Inc. All rights reserved.
```

(also applies to other files)

File: immudb/embedded/store/ongoing_tx.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from immudb.constants import *
+from immudb.printable import printable
+import immudb.embedded.store as store
+
+
+class EntrySpec(printable):
+    # def __init__(self):
```
Comment (byo) [inline immudb/embedded/store/ongoing_tx.py:20]: Please remove if not needed

File: immudb/grpc/proto/schema.proto [hunk 3]
```diff
@@ -328,38 +369,90 @@ message Database {
...
+
+	uint32 txLogCacheSize = 15;
+	uint32 vLogMaxOpenedFiles = 16;
+	uint32 txLogMaxOpenedFiles = 17;
+	uint32 commitLogMaxOpenedFiles = 18;
+
+	IndexSettings indexSettings = 19;
+}
+
+message IndexSettings {
```
Comment (byo) [inline immudb/grpc/proto/schema.proto:399]: This message was added after 1.2.2 and is now changed in current master - I think that we should use proto definition from 1.2.2 to avoid introducing breaking changes in the API.

File: immudb/printable.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class printable:
```
Comment (byo) [inline immudb/printable.py:1]: Is this file needed?

File: immudb/client.py [hunk 5]
```diff
@@ -101,6 +101,9 @@ def stub(self):
     def healthCheck(self):
         return healthcheck.call(self.__stub, self.__rs)
 
+    def health(self):
```
Comment (byo) [inline immudb/client.py:104]: There could be a confusion between health and healthCheck calls - maybe we should deprecated one and prefer the other?

File: immudb/embedded/store/kv_metadata.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    def IsExpirable(self) -> bool:
+        return EXPIRES_AT_ATTR_CODE in self.attributes
+
+    def ExpirationTime(self) -> datetime.datetime:
+        if EXPIRES_AT_ATTR_CODE in self.attributes:
+            return self.attributes[EXPIRES_AT_ATTR_CODE]
+        else:
+            raise ErrNonExpirable
+
+    # Prepared for v1.2.3
```
Comment (byo) [inline immudb/embedded/store/kv_metadata.py:64]: Let's prepare new branch with changes for v1.2.3 and put the code in there

File: immudb/handler/verifiedreference.py [hunk 7]
```diff
@@ -32,32 +33,34 @@ def call(service: schema_pb2_grpc.ImmuServiceStub, rs: RootService, refkey: byte
...
-        targetID,
-        sourceAlh,
-        targetAlh,
-    )
+    if vtx.tx.header.nentries != 1:
+        raise ErrCorruptedData
+    tx = schema.TxFromProto(vtx.tx)
+    entrySpecDigest = store.EntrySpecDigestFor(tx.header.version)
+    inclusionProof = tx.Proof(database.EncodeKey(key))
+    # TODO: Is metadata missing here?
```
Comment (byo) [inline immudb/handler/verifiedreference.py:41]: We're not setting metadata in the request (in line ~34) - thus the metadata should be empty. We should check though if it is empty before doing the proof. The proof will fail if server injected some metadata anyway but let's have a good error message here.

File: immudb/typeconv.py [hunk 8]
```diff
@@ -30,6 +34,8 @@ def sqlvalue_to_py(sqlValue):
         return sqlValue.bs
     elif sqlValue.HasField("s"):
         return sqlValue.s
+    elif sqlValue.HasField("ts"):
+        return datetime.fromtimestamp(sqlValue.ts/1e6, timezone.utc)
```
Comment (byo) [inline immudb/typeconv.py:38]: Is this division floating point? If no, then we're truncating the timestamp to second precision only.

Comment (mabmayer) [reply immudb/typeconv.py:38]: It's floating point, and the corresponding test case covers also fractions of seconds.

File: immudb/client.py (hunk 5 above)
Comment (mabmayer) [reply immudb/client.py:104]: Same in the Go SDK. Returned parameters are different.

File: immudb/client.py [hunk 9]
```diff
@@ -175,16 +178,30 @@ def databaseList(self):
         return [x.databaseName for x in dbs.dblist.databases]
 
     def databaseUse(self, dbName: bytes):
+        warnings.warn("Call to deprecated databaseUse. Use useDatabase instead",
```
[the remark this answers is not in the document]
Comment (mabmayer) [reply immudb/client.py:181]: This is up to the warnings package, and under control of the user. See https://docs.python.org/3/library/warnings.html . Default for the warnings filter: "print the first occurrence of matching warnings for each location (module + line number) where the warning is issued"

File: immudb/embedded/store/kv_metadata.py (hunk 6 above)
Comment (mabmayer) [reply immudb/embedded/store/kv_metadata.py:64]: ok

File: immudb/handler/verifiedreference.py (hunk 7 above)
Comment (mabmayer) [reply immudb/handler/verifiedreference.py:41]: hmm, the matching line in Golang is https://github.com/codenotary/immudb/blob/master/pkg/client/client.go#L1217 . The TODO was meant for clarification, if Golang SDK is doing things right here. Generally, I think we should be as close as possible to the Golang SDK.

**Comment by mabmayer** [comment]
Please visit also the Release Notes:
https://github.com/codenotary/immudb-py/releases

**Comment by Razikus** [comment]
Just refactored sql tests and added 2 new tests cases - count and transactions

**Review by byo — APPROVED** [review]
LGTM, also checked release notes - looks good
